File: 2025-01-January-GFG-POTD/January 12 - Trapping Rain Water/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxWater(vector<int>& arr) {
        int n = arr.size();
        if (n == 0) return 0;

        int left = 0, right = n - 1;
        int leftMax = 0, rightMax = 0;
        int totalWater = 0;

        while (left <= right) {
            if (arr[left] <= arr[right]) {
                if (arr[left] >= leftMax) {
                    leftMax = arr[left];
                } else {
                    totalWater += leftMax - arr[left];
                }
                left++;
            } else {
                if (arr[right] >= rightMax) {
                    rightMax = arr[right];
                } else {
                    totalWater += rightMax - arr[right];
                }
                right--;
            }
        }

        return totalWater;
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin.ignore(); // To handle newline after integer input
        string input;
        getline(cin, input);

        vector<int> arr;
        stringstream ss(input);
        string temp;
        while (getline(ss, temp, ' ')) {
            arr.push_back(stoi(temp));
        }

        Solution obj;
        cout << obj.maxWater(arr) << endl;
        cout << "~" << endl;
    }
    return 0;
}

File: 2025-01-January-GFG-POTD/January 12 - Trapping Rain Water/Solution.py
class Solution:
    def maxWater(self, arr):
        n = len(arr)
        if n == 0:
            return 0

        left, right = 0, n - 1
        left_max, right_max = 0, 0
        total_water = 0

        while left <= right:
            if arr[left] <= arr[right]:
                if arr[left] >= left_max:
                    left_max = arr[left]
                else:
                    total_water += left_max - arr[left]
                left += 1
            else:
                if arr[right] >= right_max:
                    right_max = arr[right]
                else:
                    total_water += right_max - arr[right]
                right -= 1

        return total_water


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        obj = Solution()
        print(obj.maxWater(arr))
        print("~")
